src/oscar/apps/catalogue/documents.py

```diff
@@ -0,0 +1,226 @@
+from dateutil.relativedelta import relativedelta
+
+from django.conf import settings
+from django.core.exceptions import ObjectDoesNotExist
+from django.db import ProgrammingError, OperationalError
+from django.db.models import QuerySet
+from django.utils import timezone
+from django.utils.html import strip_tags
+from django.utils.six import with_metaclass
+
+from oscar.apps.catalogue.models import AttributeOption
+from oscar.core.loading import get_model
+
+from django_elasticsearch_dsl.documents import DocTypeMeta
+from django_elasticsearch_dsl import DocType, fields
+from elasticsearch_dsl import Mapping, MetaField
+
+
+Line = get_model('order', 'Line')
+Product = get_model('catalogue', 'Product')
+ProductAttribute = get_model('catalogue', 'ProductAttribute')
+StockRecord = get_model('partner', 'StockRecord')
+
+product_mapping = Mapping('product')
+product_mapping.field('raw_title', 'text', boost=1.25)
+product_mapping.field('all_skus', 'text', analyzer='standard')
+
+
+ATTRIBUTE_TYPE_ES_FIELDS = {
+    ProductAttribute.TEXT: fields.KeywordField,
+    ProductAttribute.INTEGER: fields.IntegerField,
+    ProductAttribute.BOOLEAN: fields.BooleanField,
+    ProductAttribute.FLOAT: fields.FloatField,
+    ProductAttribute.RICHTEXT: fields.KeywordField,
+    ProductAttribute.DATE: fields.DateField,
+    ProductAttribute.DATETIME: fields.DateField,
+    ProductAttribute.OPTION: fields.KeywordField,
+    ProductAttribute.MULTI_OPTION: fields.KeywordField
+}
+
+ES_FIELDS = {
+    'attachment': fields.AttachmentField,
+    'boolean': fields.BooleanField,
+    'byte': fields.ByteField,
+    'completion': fields.CompletionField,
+    'date': fields.DateField,
+    'double': fields.DoubleField,
+    'float': fields.FloatField,
+    'geo_point': fields.GeoPointField,
+    'geo_shape': fields.GeoShapeField,
+    'integer': fields.IntegerField,
+    'ip': fields.IpField,
+    'keyword': fields.KeywordField,
+    'long': fields.LongField,
+    'nested': fields.NestedField,
+    'object': fields.ObjectField,
+    'short': fields.ShortField,
+    'string': fields.StringField,
+    'text': fields.TextField
+}
+
+
+class ProductDocumentMeta(DocTypeMeta):
+
+    def __new__(cls, name, bases, attrs):
+        attrs['product_attributes'] = []
+
+        try:
+            attribute_configs = settings.OSCAR_SEARCH['INDEXED_FIELDS'].get('product', {})
+            indexed_attributes = ProductAttribute.objects.filter(code__in=attribute_configs.keys())
+
```

Comment:
It seems that this construct hints at an incompatibility between elasticsearch-dsl and oscar's datascructure.
If we need to make a query to create new types/classes - which are later converted **back** to json (elasticsearch query language) - it would follow it is better to _skip_ the creation of types entirely and revert back to using json directly. Looks like the extra abstraction layer elasticsearch-dsl introduces is more of a burden than a benefit.